bulletroboy/exoforce/exoforce_hardware.py

```diff
@@ -0,0 +1,266 @@
+		if choice in YES_OPTIONS:
+			self.calibration_timer = self.create_timer(2, self.calibration)
+		else:
+			self.start_exoforce()
+
+		self.applying_force = False
+		self.applying_force_timer = self.create_timer(1, self.stop_applying_force)
+		self.applying_force_timer.cancel()
+
+	def stop_applying_force(self):
+		self.applying_force = False
+
+	def start_exoforce(self):
+		self.create_timer(0.5, self.update)
+		choice = input("\nInitialize position control? (Yes/no): ")
+		if choice in YES_OPTIONS:
+			self.init_pos_control()
+
+	def init_tendons(self):
+		choice = input("\nInitialize tendons? (Yes/no): ")
+		if choice in YES_OPTIONS:
+			for ef in self.end_effectors:
+				input("\n- Press enter to initialize " + ef + " tendons...")
+				motor_ids = []
+				set_points = []
+				for muscle in self.get_ef_muscle_units(ef):
+					motor_ids.append(muscle.id)
+					set_points.append(float(FAST_PWM_TENSION * muscle.direction))
+				self.send_motor_commands(motor_ids, set_points)
+				input("- Press enter to finish")
+				set_points = []
+				for muscle in self.get_ef_muscle_units(ef):
+					set_points.append(float(MIN_PWM_TENSION * muscle.direction))
+				self.send_motor_commands(motor_ids, set_points)
+			self.send_motor_commands([16],[float(MIN_PWM_TENSION)])
+			print("\nTendons ready!\n")
+	
+	# def init_tendons(self):
+	# 	choice = input("\nInitialize tendons? (Yes/no): ")
+	# 	if choice in YES_OPTIONS:
+	# 		print("\nInitiating tendons:")
+	# 		for muscle in self.muscle_units:
+	# 			tendon = muscle.id
+	# 			print("\n- Tendon "+ str(tendon) + ":")
+	# 			input("Press enter to start...")
+	# 			self.send_motor_commands([tendon], [float(FAST_PWM_TENSION * muscle.direction)])
+	# 			input("Press enter to slowdown...")
+	# 			self.send_motor_commands([tendon], [float(MIN_PWM_TENSION * muscle.direction)])
+	# 			# input("Press enter to stop...")
+	# 			# self.send_motor_commands([tendon], [0.0])
+	# 		print("\nTendons ready!\n")
+
+	def calibration(self):
+		print(self.end_effectors.values())
+		if np.any([ef["position"] is None for ef in self.end_effectors.values()]):
+			print("Waiting for end effector initial poses...")
+			return
+
+		print("\nInitializing motor position calibration:")
+		for muscle in self.muscle_units:
+			input("- Touch attachment point " + str(muscle.id) + " with the right hand tracker and press enter...")
+			muscle.motor.via_point.world_point = self.end_effectors["right_wrist"]["position"]
+			print("Motor " + str(muscle.id) + " position updated to: " + str(muscle.motor.via_point.world_point))
+
+		print("\nMotor positions updated!\n")
+		self.start_exoforce()
+		self.calibration_timer.destroy()
+
+
+	def print_distance(self):
+		for muscle in self.muscle_units:
+			dist = np.linalg.norm(muscle.motor.via_point.world_point - self.end_effectors["right_wrist"]["position"])
+			if dist < 0.1:
+				self.get_logger().info("Motor " + str(muscle.id) + " touched!")
+			if muscle.id == 3:
```

Comment:
what is 3?

Reply:
see https://github.com/Roboy/bulletroboy/pull/25#issuecomment-760128816